HeterogeneousCore/AlpakaCore/interface/CopyToDeviceCache.h

```diff
@@ -0,0 +1,111 @@
+    //
+    // Doing with template specialization (rather than
+    // std::conditional_t and if constexpr) because the
+    // CopyToDevice<THostObject>::copyAsync() is ill-defined e.g. for
+    // PortableCollection on host device
+    template <typename TDev, typename TQueue, typename THostObject>
+    class CopyToDeviceCacheImpl {
+    public:
+      using Device = TDev;
+      using Queue = TQueue;
+      using HostObject = THostObject;
+      using Copy = CopyToDevice<HostObject>;
+      using DeviceObject = decltype(Copy::copyAsync(std::declval<Queue&>(), std::declval<HostObject const&>()));
+
+      CopyToDeviceCacheImpl(HostObject const& srcObject) {
+        using Platform = alpaka::Platform<Device>;
+        auto const& devices = cms::alpakatools::devices<Platform>();
+        std::vector<std::shared_ptr<Queue>> queues;
+        queues.reserve(devices.size());
+        data_.reserve(devices.size());
+        for (auto const& dev : devices) {
+          auto queue = getQueueCache<Queue>().get(dev);
+          data_.emplace_back(Copy::copyAsync(*queue, srcObject));
+          queues.emplace_back(std::move(queue));
+        }
+        for (auto& queuePtr : queues) {
+          alpaka::wait(*queuePtr);
+        }
+      }
+
+      DeviceObject const& get(size_t i) const { return data_[i]; }
+
+    private:
+      std::vector<DeviceObject> data_;
+    };
+
+    // For host device, copy the host object directly instead
+    template <typename TQueue, typename THostObject>
+    class CopyToDeviceCacheImpl<alpaka_common::DevHost, TQueue, THostObject> {
+    public:
+      using HostObject = THostObject;
+      using DeviceObject = HostObject;
+
+      CopyToDeviceCacheImpl(HostObject const& srcObject) : data_(srcObject) {}
+
+      DeviceObject const& get(size_t i) const { return data_; }
+
+    private:
+      HostObject data_;
+    };
+  }  // namespace detail
+
+  /**
+   * This class template implements a cache for data that is copied
+   * from the host (of type THostObject) to all the devices
+   * corresponding the TQueue queue type.
+   *
+   * The host-side object to be copied is given as an argument to the
+   * class constructor. The constructor uses the
+   * CopyToDevice<THostObject> class template to perfom the copy, and
+   * waits for the data copies to finish, i.e. the constructor is
+   * synchronous wrt. the data copies.
+   *
+   * The device-side object corresponding to the THostObject (actual
+   * type is the return type of CopyToDevice<THostObject>::copyAsync())
+   * can be obtained with get() member function, that has either the
+   * queue or device argument.
+   *
+   * TODO: In principle it would be better to template over Device,
+   * but then we'd need a way to have a "default queue" type for each
+   * Device in order to infer the return type of
+   * CopyToDevice::copyAsync(). Alternatively, the template over
+   * TQueue could be removed by moving the class definition to
+   * ALPAKA_ACCELERATOR_NAMESPACE.
```

Comment:
Another option could be to use
  - synchronous queues for the CPU backends (serial, tbb, openmp, etc.)
  - asynchronous queues for the GPU backends (cuda, rocm, sycl)

If we ever implement support for the "parallel" CPU backends (tbb, openmp blocks) we would likely use async queues - but here it would make sense to use a sync queue anyway, because there is only once CPU device, CPU don't really benefit much from being done in parallel, and we synchronise at the end anyway.

In fact, since the host-to-host copies are done by the specialisation, we could just always assume async queues.

The SYCL CPU backend would likely still use an async queue (because it has a different platform and device type) but that would be the case anyway unless we add an explicit specialisation for it.

Reply:
> In fact, since the host-to-host copies are done by the specialisation, we could just always assume async queues.

Thanks, this was a good point. I decided to make an attempt to define a mapping for Device to Queue types (for non-host Devices) in order to have "more correct" interface now, and it didn't seem too bad. Let me know what you think.